Subject: [PATCH] git-remote-mediawiki: avoid redundant imports

Previously the git-remote mechanism would request an import of HEAD
and then an import of whatever it pointed to.

In that case git-remote-mediawiki simply ignored the request for HEAD
and proceeded with the following request.

However, later versions of Git are now resolving the HEAD symbolic ref
before issuing the import commands resulting in two import commands
for the same ref in a row.  This reesults in an unwanted doubling of
the history and the import taking twice as long.

Correct this by avoiding multiple imports for the same ref.  We currently
only detect this when the import commands are grouped together, but that
seems to always be the case when a symbolic ref is resolved.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 contrib/mw-to-git/git-remote-mediawiki.perl | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/contrib/mw-to-git/git-remote-mediawiki.perl b/contrib/mw-to-git/git-remote-mediawiki.perl
index 41e74fba..e4603da7 100755
--- a/contrib/mw-to-git/git-remote-mediawiki.perl
+++ b/contrib/mw-to-git/git-remote-mediawiki.perl
@@ -772,7 +772,10 @@ sub get_more_refs {
 sub mw_import {
 	# multiple import commands can follow each other.
 	my @refs = (shift, get_more_refs('import'));
+	my %seen = ();
 	foreach my $ref (@refs) {
+		next if $seen{$ref};
+		$seen{$ref} = 1;
 		mw_import_ref($ref);
 	}
 	print {*STDOUT} "done\n";
---
